src/main/java/com/algorand/algosdk/transaction/AppBoxReference.java

```diff
@@ -0,0 +1,55 @@
+package com.algorand.algosdk.transaction;
+
+import com.algorand.algosdk.util.BoxQueryEncoding;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public class AppBoxReference {
+    // the app ID of the app this box belongs to. Instead of serializing this value,
+    // it's used to calculate the appIdx for AppBoxReference.
+    private final long appId;
+
+    // the name of the box unique to the app it belongs to
+    private final byte[] name;
+
+    public AppBoxReference(long appId, byte[] name) {
+        this.appId = appId;
+        this.name = Arrays.copyOf(name, name.length);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        AppBoxReference that = (AppBoxReference) o;
+        return appId == that.appId && Arrays.equals(name, that.name);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(appId);
+        result = 31 * result + Arrays.hashCode(name);
+        return result;
```

Comment:
nit: @jasonpaulos we had a short discussion on this one before, if I remembered correctly. We introduced `ComparableBytes` to handle such problem, right? Do we want to change implementation on this one?

Reply:
This is a separate concern from `ComparableBytes`.

Previously the code made a `HashSet` from `AppBoxReference` instances, so `hashCode` was required to be implemented. Now that that is no longer the case, we technically don't need `hashCode`, but I think there's no harm in keeping it.

@michaeldiamant likely has more insight into this than I do.

Reply:
@ahangsu Thoughts on my end - Let me know if the brings enough clarity to resolve?  If not, happy to discuss live.
* I agree with @jasonpaulos's assessment that the highlighted `hashCode` override is largely orthogonal to `ComparableBytes`.
* For a class holding data (like `AppBoxReference`), I think it's a bug if `hashCode` + `equals` are _not_ overwritten.  The class _is_ intended to be used for comparison purposes.  So, I think it's correct to leave as is.
  * In Scala, the kind of class we're describing is a _case class_:  https://docs.scala-lang.org/tour/case-classes.html.    
  * As of Java 14, Java introduces the concept as _record classes_: 
 https://docs.oracle.com/en/java/javase/16/language/records.html
  * I realize the repo convention is to _omit_ `hashCode`.  Based on the above rationale, I view it as a bug:  https://github.com/algorand/java-algorand-sdk/issues/344.

Reply:
I see, my apologies for wronged memory. I was somewhat concerned about the implementation of the `hashcode` but now I am more clear about the situation.

I guess the most straightforward way to get around explicit definition of hashcode is to up the jvm version, at some point in the future...

Reply:
@ahangsu No need to apologize, it's a fair question.  For reference, I use the IDE (intellij) to auto-generate `equals` and `hashCode` implementations.  I find allowing the IDE to generate these methods is less error prone than a handwritten version.